### PR TITLE
tox.ini: configure and run tox

### PR DESCRIPTION
**Pull request by PrajwalM2212**
This PR adds tox.ini file and modifies
pull_request.yml so that github actions
can run tox.

Signed-off-by: PrajwalM2212 <prajwalmmath@gmail.com>

### CODE REVIEW

**Comment by PrajwalM2212** [comment]
I will be deliberately making the PR to fail to see if it is caught by tox.

**Comment by PrajwalM2212** [comment]
I have got `Test Changes` and `Test Coverage` to work on tox. 3 more to go.

**Comment by PrajwalM2212** [comment]
@nishakm @rnjudge 
I have added `tox`, the recommended way for testing . This took a lot of effort to complete. I even cloned the tox repo and had to read it some extent to understand how it works.

File: tern/analyze/common.py [hunk 1]
```diff
@@ -164,6 +164,20 @@ def convert_to_pkg_dicts(pkg_dict):
...
+
+
+
+
+
+
+
+
+
+
```
Comment (nishakm) [inline tern/analyze/common.py]: I think this was unintentional?

Comment (PrajwalM2212) [reply tern/analyze/common.py]: Did that to check if prospector in tox caches it.

**Comment by PrajwalM2212** [comment]
I was reading the docs and found that we can pass around environment variables. Have used that to fix the problem.

### Prospector failing example
Screen Shot 2020-02-22 at 8 48 50 PM
Screen Shot 2020-02-22 at 8 48 27 PM

### Bandit running example
Screen Shot 2020-02-22 at 8 48 40 PM

So tox is functioning correctly.

**Comment by nishakm** [comment]
@PrajwalM2212 Thanks again for all the hard work! I think this ended up being more complicated than it should have :). The tests in the `ci` directory are for the project's CI and not for regular development. So we don't have to replace the CircleCI or the GHA configurations.
However, we do need tox (as you had filed in #406) to help developers before they submit the PR. For this purpose, just running `python tests/...` for each of the tests would be enough. I still don't understand how tox would behave from within a virtualenv. I'll check out the behaviour. Another place we would need tox is for deployments. In this case we will be running the whole battery of tests rather than just what the PR changed. But we can get to that part later. For now, just running all the tests in `tests` will suffice.

**Comment by nishakm** [comment]
@PrajwalM2212 I tested with a simple `tox.ini` file:
```
[tox]
envlist = py35, py36, py37

[testenv]
deps =
    bandit~=1.6
    GitPython~=2.1
    prospector>=1.2
commands =
    python -m unittest discover -s tests
```
This works within the virtualenv with errors that python interpreters were not found:
```
ERROR:  py35: InterpreterNotFound: python3.5
  py36: commands succeeded
  py37: commands succeeded
```
I think this is OK.

File: tox.ini [hunk 2]
```diff
@@ -0,0 +1,38 @@
+[tox]
+envlist = commit
+          prospector
+          security
+          changes
+          coverage
```
Comment (nishakm) [inline tox.ini]: Would it be possible to run this on multiple python environments?
```
envlist = py35, py36, py37
```

File: tox.ini [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+[testenv]
+description = Run tests
+basepython = python3.6
+deps = GitPython~=2.1
+passenv = FILES_TO_LINT
```
Comment (nishakm) [inline tox.ini]: I don't think we need to pass environment variables here.

File: tox.ini [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+[testenv:prospector]
+description = Run the Prospector test
+deps = prospector>=1.1
```
Comment (nishakm) [inline tox.ini]: prospector>=1.2

File: tox.ini [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+commands = /bin/sh -c 'if [ -z $FILES_TO_LINT ]; then echo "No .py files to lint"; else echo $FILES_TO_LINT | xargs prospector; fi'
+           /bin/sh -c 'echo $FILES_TO_LINT'
```
Comment (nishakm) [inline tox.ini]: Just run `prospector` here

File: tox.ini [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+[testenv:commit]
+description = Run the Commit test
+commands = python ci/test_commit_message.py
+
+[testenv:security]
+description = Run bandit
+deps = bandit~=1.6
+commands = /bin/sh -c 'if [ -z $FILES_TO_LINT ]; then echo "No .py files to lint"; else echo $FILES_TO_LINT | xargs bandit; fi'
```
Comment (nishakm) [inline tox.ini]: Just run `bandit` here.

File: tox.ini [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+[testenv:changes]
+description = Run changes
+commands = python ci/test_files_touched.py
+
```
Comment (nishakm) [inline tox.ini]: Just run tests here: `python -m unittest discover -s tests`

File: tox.ini [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+[testenv:coverage]
+description = Get current coverage
+deps = coverage
+       pytest
+commands = coverage run -m pytest tests
+           coverage report --omit .tox/**,tests/**
```
Comment (nishakm) [inline tox.ini]: No need to add the `pytest` dependency. You can run `coverage run -m unittest discover -s tests`.

File: tox.ini (hunk 6 above)
Comment (nishakm) [reply tox.ini]: `commands = bandit .`

**Comment by PrajwalM2212** [comment]
@nishakm Changes done. Tests can now be run easily. It is really nice to be able to run these tests before submitting the PR.

File: setup.cfg [hunk 9]
```diff
@@ -54,4 +54,4 @@ console_scripts =
     tern = tern.__main__:main
 
 [options.extras_require]
-dev = bandit~=1.6; prospector>=1.2; GitPython~=2.1
+dev = bandit~=1.6; prospector>=1.2; GitPython~=2.1; tox
```
Comment (nishakm) [inline setup.cfg]: On my machine, the tox version is 3.14.5 Is that the same for you? If yes, then add `tox>=3.14`

File: tox.ini [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+[testenv:commit]
+description = Run the Commit test
+commands = python ci/test_commit_message.py
+
+[testenv:security]
+description = Run bandit
+deps = bandit~=1.6
+commands = bandit .
+
+[testenv:changes]
```
Comment (nishakm) [inline tox.ini]: Change to `tests`

**Comment by PrajwalM2212** [comment]
@nishakm Done

**Pull request merged**